### PR TITLE
misc: making some internal api types more consistent

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Our internal ch4 in some place uses inconsistent types for the same parameters, in some place parameters with same name refers to different semantics. The inconsistency can be confusing and in some place places unnecessary hidden hazard.

This PR fixes most of these inconsistencies. 

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
@minsii Since you mentioned some of the internal type inconsistency, would you like to review this PR? It is a set of internal API type changes to make them more consistent and get ready for the large count API in MPI 4.

**Comment by minsii** [comment]
Will do

File: src/mpid/ch4/src/ch4r_callbacks.c [hunk 1]
```diff
@@ -549,7 +549,7 @@ int MPIDIG_send_cts_target_msg_cb(int handler_id, void *am_hdr, void *data, MPI_
         mpi_errno =
             MPIDI_SHM_am_isend_reply(MPIDIG_REQUEST(sreq, req->sreq).context_id,
                                      MPIDIG_REQUEST(sreq, rank), MPIDIG_SEND_DATA,
-                                     &send_hdr, sizeof(send_hdr),
+                                     &send_hdr, (MPI_Aint) sizeof(send_hdr),
                                      MPIDIG_REQUEST(sreq, req->sreq).src_buf,
```
Comment (minsii) [inline src/mpid/ch4/src/ch4r_callbacks.c:553]: As we are always casting return of `sizeof`, do you think it might benefit if we define a utility macro for it? For example:
```
#define MPIR_SIZEOF(var) (MPI_Aint) sizeof((var))
```

File: src/mpid/ch4/ch4_api.txt [hunk 2]
```diff
@@ -26,8 +26,8 @@ Non Native API:
       NM : comm_ptr
      SHM : comm_ptr
   mpi_open_port : int
-      NM : info_ptr, port_name-2
-     SHM : info_ptr, port_name-2
+      NM : info, port_name-2
+     SHM : info, port_name-2
   mpi_close_port : int
```
Comment (minsii) [inline src/mpid/ch4/ch4_api.txt:31]: Change of input pointer parameter is not described in the commit message. Maybe better to separate to a different commit.

**Review by minsii — COMMENTED** [review]
Looks OK to me. Just a few comments.

File: src/mpid/ch4/src/ch4r_callbacks.c (hunk 1 above)
Comment (hzhou) [reply src/mpid/ch4/src/ch4r_callbacks.c:553]: I think it is nice to have. But I am also aware there is another desire to reduce the macro usage. I think this is a case with marginal benefit and someone may even be against hiding the cast (another evil thing) behind a macro.

File: src/mpid/ch4/ch4_api.txt (hunk 2 above)
Comment (hzhou) [reply src/mpid/ch4/ch4_api.txt:31]: Will do

Comment (hzhou) [reply src/mpid/ch4/ch4_api.txt:31]: Done.

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Review by minsii — APPROVED** [review]

**Comment by wesbland** [comment]
I'm pretty sure this is moving in the wrong direction. Look at Section 2.5.6 and 2.5.8 in the upcoming MPI 4.0 standard (these are incomplete excerpts and I'd recommend reading the whole thing to get the full context):

### MPI_Aint

> Some MPI procedures use address arguments that represent an absolute address in the call- ing program, or relative displacement arguments that represent differences of two absolute addresses. The datatype of such arguments is MPI_Aint in C and INTEGER(KIND= MPI_ADDRESS_KIND) in Fortran. These types must have the same width and encode address values in the same manner such that address values in one language may be passed directly to another language without conversion.

### MPI_Count 

> At times, one needs a single type that can be used to address locations within either memory or files as well as express count values, and that type is MPI_Count in C and INTEGER(KIND=MPI_COUNT_KIND) in Fortran.
...
Even though the MPI_Count type is large enough to encode address locations, the MPI_Count type shall not be used to represent an absolute address.

**Comment by hzhou** [comment]
> I'm pretty sure this is moving in the wrong direction.

What direction are you referring to? Or which change in this PR are you questioning?

**Comment by wesbland** [comment]
> > I'm pretty sure this is moving in the wrong direction.
> 
> @wesbland What direction are you referring to? Or which change in this PR are you questioning?

This PR seems to be moving things from `MPI_Count` to `MPI_Aint`. My understanding is that `MPI_Aint` is theoretically a vector type (as in it could be an address for different parts of memory) while `MPI_Count` is basically just a big integer.

It probably doesn't make a lot of difference most of the time, but it could cause problems in the future.

**Comment by hzhou** [comment]
At user-level or MPI-layer, we are strictly adhering to the standard, and we are differentiating between `MPI_Count` and `MPI_Aint`. However, internally, we are not using `MPI_Count` directly. This is for 1, because `MPI_Count` potentially can be larger than `MPI_Aint` and internally use it may incur bigger performance impact than what we are comfortable. And for 2, a bigger integer than address space doesn't make sense in most of the cases.

The direction is to use `MPI_Aint` throughout *internally* and *confine* the `MPI_Count` treatment to the glue layer between MPI-layer and internal. This "glue" layer may be a simple assertion for the case of `MPI_Count` input and a larger than `MPI_Aint` value doesn't make practical sense, or we'll have special handling routines which will make use of explicit struct for larger than `MPI_Aint` values -- similar to what do with `MPI_Status`. I think it is simply too dangerous to allow implicit handling/conversion for cases where the semantic difference between `MPI_Count` and `MPI_Aint` matters.
